Dark mode does not work in SVGImages

shallawa opened the pull request:
#### 702c47ccd9244a4f95caddb607cdb2b638c0fa77
<pre>
Dark mode does not work in SVGImages
<a href="https://bugs.webkit.org/show_bug.cgi?id=283489">https://bugs.webkit.org/show_bug.cgi?id=283489</a>
<a href="https://rdar.apple.com/140661763">rdar://140661763</a>

Reviewed by Simon Fraser.

Page::useColorAppearance() is not propagated to the Page of the SVGImage when it
is first created. To fix this issue, SVGImage will query the container Page
useSystemDarkAppearance() through the ImageObserver and set it in its Page.

Also if the dark mode changes after the SVGImage is loaded, the SVGImage does not
recalculate its style colors. To sync the setUseColorAppearance() of all the
SVGImages when the system appearance changes, Page::setUseColorAppearance() needs
to scan all the SVGImages in all Documents and call setUseColorAppearance() of
the Pages of these Documents.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/svg/custom/resources/light-dark-red-green.svg: Added.
* LayoutTests/svg/custom/resources/prefers-color-scheme-dark-red-green.svg: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-dynamic-expected.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-dynamic.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-initial-expected.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-initial.html: Added.
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::useSystemDarkAppearance const):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::useSystemDarkAppearance const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUseColorAppearance):
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::CachedImageListener::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::useSystemDarkAppearance const):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::dataChanged):

Canonical link: <a href="https://commits.webkit.org/303920@main">https://commits.webkit.org/303920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4cd913e24bcc5b771cc9ea5191c93f7bf9dd719b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133934 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6445 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141514 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85995 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e80d2069-a4d5-4ef7-89ed-3cf73b0102f1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6309 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102469 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120069 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83266 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/444050dd-ef27-450c-a894-127de42215a4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4780 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2400 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144159 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6115 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110815 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6197 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111023 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28168 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4637 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116327 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59862 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6167 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34585 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6013 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69631 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6258 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6121 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/loader/cache/CachedImage.cpp:
```diff
@@ -743,6 +743,14 @@ void CachedImage::scheduleRenderingUpdate(const Image& image)
         client->scheduleRenderingUpdateForImage(*this);
 }
 
+bool CachedImage::useSystemDarkAppearance()
```
Can this be `const` ?

shallawa replied:
Yes it can be. Fixed.

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1712,6 +1712,13 @@ bool RenderElement::allowsAnimation() const
 
+bool RenderElement::useSystemDarkAppearance() const
+{
+    if (RefPtr page = document().page())
```
Do we need to consult style here, like `computeTextPaintStyle()` does?

shallawa replied:
I do not think we need to do that. All we need is to transfer the system dark appearance to the SVGImage. I do not think the style of the RenderImage which references the SVGImage should affect the styles inside the SVGImage.

smfr commented on LayoutTests/svg/custom/svg-image-dark-mode-dynamic.html:
```diff
@@ -0,0 +1,46 @@
+    <table>
+        <tr>
+            <td>
+                <svg style="color-scheme: light dark;" fill="light-dark(red, green)">
```
Does `light-dark` as a valid function here just work because fill is a CSS presentation attribute?

Please also test `light-dark()` on an element inside the SVG, and `@media (prefers-color-scheme: dark)` on CSS inside the SVG.

shallawa committed:
Dark mode does not work in SVGImages
https://bugs.webkit.org/show_bug.cgi?id=283489
rdar://140661763

Reviewed by Simon Fraser.

Page::useColorAppearance() is not propagated to the Page of the SVGImage when it
is first created. To fix this issue, SVGImage will query the container Page
useSystemDarkAppearance() through the ImageObserver and set it in its Page.

Also if the dark mode changes after the SVGImage is loaded, the SVGImage does not
recalculate its style colors. To sync the setUseColorAppearance() of all the
SVGImages when the system appearance changes, Page::setUseColorAppearance() needs
to scan all the SVGImages in all Documents and call setUseColorAppearance() of
the Pages of these Documents.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/svg/custom/resources/light-dark-red-green.svg: Added.
* LayoutTests/svg/custom/resources/prefers-color-scheme-dark-red-green.svg: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-dynamic-expected.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-dynamic.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-initial-expected.html: Added.
* LayoutTests/svg/custom/svg-image-dark-mode-initial.html: Added.
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::useSystemDarkAppearance const):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/loader/cache/CachedImageClient.h:
(WebCore::CachedImageClient::useSystemDarkAppearance const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUseColorAppearance):
* Source/WebCore/platform/graphics/ImageObserver.h:
(WebCore::ImageObserver::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::CachedImageListener::useSystemDarkAppearance const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::useSystemDarkAppearance const):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::dataChanged):

Canonical link: https://commits.webkit.org/303920@main
